From 3d1b45f4d15c6785f93fa42ad909f853f37a19eb Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Tue, 13 Sep 2022 17:31:35 +0200
Subject: [PATCH 215/954] fix(st-ddr): phy: add phyvref and sequencectrl
 advanced parameters

Replace hard-coded values by user input advanced structure fields. They
can be generated during configuration setting.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git c783c1b3375c4a44062ee580763d43e88873d6cb fix(st-ddr): phy: add phyvref and sequencectrl advanced parameters]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I6bf21af257e761f0514daaf0f04e370eead59e4c
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/266085
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 .../phyinit/include/ddrphy_phyinit_struct.h   | 70 ++++++++++++++++---
 .../phyinit/src/ddrphy_phyinit_initstruct.c   | 29 +-------
 2 files changed, 61 insertions(+), 38 deletions(-)

diff --git a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h
index 8e70d808f2..f381a269e1 100644
--- a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h
+++ b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_struct.h
@@ -968,18 +968,66 @@ typedef struct user_input_advanced {
 					 *   0x0 | Minimizes number of Imem/Dmem loads (default)
 					 */
 
-	int enabledficspolarityfix;     /* Enable alternative PIE program */
+	int enabledficspolarityfix;     /*
+					 * Enable alternative PIE program
+					 *
+					 * See STAR 9001524249 for details on this workaround. Set
+					 * to 1 if PUB_VERSION <2.43a, otherwise set to 0. If
+					 * enabled the PIE programs Dfi{Rd,Wr}DataCsDestMap CSR's
+					 * to default values 0x00E4 before running PPT.
+					 * Before exiting PPT, PIE will restore
+					 * Dfi{Rd,Wr}DataCsDestMap CSR's to 0x00E1.
+					 *
+					 * Value | Description
+					 * ----- | ---
+					 *   0x0 | Disable (default)
+					 */
 
-	/*
-	 * See STAR 9001524249 for details on this workaround. Set to 1 if
-	 * PUB_VERSION <2.43a, otherwise set to 0. If enabled the PIE programs
-	 * Dfi{Rd,Wr}DataCsDestMap CSR's to default values 0x00E4 before runnning PPT.
-	 * Before exiting PPT, PIE will restore Dfi{Rd,Wr}DataCsDestMap CSR's to 0x00E1.
-	 *
-	 * Value | Description
-	 * ----- | ---
-	 *   0x0 | Disable (default)
-	 */
+	int phyvref;			/*
+					 * Must be programmed with the Vref level to be used by the
+					 * PHY during reads
+					 *
+					 * The units of this field are a percentage of VDDQ
+					 * according to the following equation:
+					 *
+					 * Receiver Vref = VDDQ*phyvref[6:0]/128
+					 *
+					 * For example to set Vref at 0.75*VDDQ, set this field to
+					 * 0x60.
+					 *
+					 * For digital simulation, any legal value can be used. For
+					 * silicon, the users must calculate the analytical Vref by
+					 * using the impedances, terminations, and series resistance
+					 * present in the system.
+					 */
+
+	int sequencectrl[NB_PS];	/*
+					 * Controls the training steps to be run. Each bit
+					 * corresponds to a training step.
+					 *
+					 * If the bit is set to 1, the training step will run.
+					 * If the bit is set to 0, the training step will be
+					 * skipped.
+					 *
+					 * Training step to bit mapping:
+					 * sequencectrl[0] = Run DevInit - Device/phy
+					 *		     initialization. Should always be set.
+					 * sequencectrl[1] = Run WrLvl - Write leveling
+					 * sequencectrl[2] = Run RxEn - Read gate training
+					 * sequencectrl[3] = Run RdDQS1D - 1d read dqs training
+					 * sequencectrl[4] = Run WrDQ1D - 1d write dq training
+					 * sequencectrl[5] = RFU, must be zero
+					 * sequencectrl[6] = RFU, must be zero
+					 * sequencectrl[7] = RFU, must be zero
+					 * sequencectrl[8] = Run RdDeskew - Per lane read dq deskew
+					 *		     training
+					 * sequencectrl[9] = Run MxRdLat - Max read latency training
+					 * sequencectrl[10] = RFU, must be zero
+					 * sequencectrl[11] = RFU, must be zero
+					 * sequencectrl[12] = RFU, must be zero
+					 * sequencectrl[13] = RFU, must be zero
+					 * sequencectrl[15-14] = RFU, must be zero
+					 */
 } user_input_advanced_t;
 
 /*
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_initstruct.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_initstruct.c
index 91865383bb..e4cef3a928 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_initstruct.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_initstruct.c
@@ -74,17 +74,6 @@ void ddrphy_phyinit_initstruct(void)
 					 * Set cspresent[3]   = 1 (if CS3 is populated with DRAM)
 					 * Set cspresent[7:4] = 0 (Reserved; must be set to 0)
 					 */
-#if STM32MP_DDR3_TYPE
-	uint8_t phyvref = 0x40U;	/*
-					 * Use Analytical VREF and Compensate for T28 Attenuator,
-					 * see PHY databook
-					 */
-#else
-	uint8_t phyvref = 0x5EU;	/*
-					 * Use Analytical VREF and Compensate for T28 Attenuator,
-					 * see PHY databook
-					 */
-#endif /* STM32MP_DDR3_TYPE */
 	uint8_t dfimrlmargin = 0x01U;	/* 1 is typically good in DDR3 */
 #if STM32MP_DDR3_TYPE
 	uint8_t addrmirror = 0x00U;	/*
@@ -160,7 +149,6 @@ void ddrphy_phyinit_initstruct(void)
 					 */
 #endif /* STM32MP_DDR3_TYPE */
 #elif STM32MP_LPDDR4_TYPE
-	uint8_t phyvref = 0x14U;
 	uint8_t caterminatingrankcha = 0x00U; /*Usually Rank0 is terminating rank */
 	uint8_t caterminatingrankchb = 0x00U; /*Usually Rank0 is terminating rank */
 	uint8_t dfimrlmargin = 0x02U; /*This needs to be large enough for max tDQSCK variation */
@@ -193,8 +181,6 @@ void ddrphy_phyinit_initstruct(void)
 	 * ##############################################################
 	 */
 
-	uint16_t sequencectrl[NB_PS];
-
 #if STM32MP_DDR4_TYPE
 	uint16_t alt_cas_l[NB_PS]; /* Need to set if using RDDBI */
 	uint16_t alt_wcas_l[NB_PS]; /* Need to set if using 2tck Write Preambles */
@@ -210,17 +196,6 @@ void ddrphy_phyinit_initstruct(void)
 
 	VERBOSE("%s Start\n", __func__);
 
-	for (myps = 0; myps < NB_PS; myps++) {
-		if (myps == 0) {
-			sequencectrl[0] = 0x031FU;
-		} else {
-			sequencectrl[myps] = 0x021FU;
-		}
-#if STM32MP_LPDDR4_TYPE
-		sequencectrl[myps] |= 0x1000U;
-#endif /* STM32MP_LPDDR4_TYPE */
-	}
-
 #if STM32MP_DDR4_TYPE
 	for (myps = 0; myps < NB_PS; myps++) {
 		alt_cas_l[myps] = 0x0000U;
@@ -261,13 +236,13 @@ void ddrphy_phyinit_initstruct(void)
 	/* 1D message block defaults */
 	for (myps = 0; myps < NB_PS; myps++) {
 		mb_ddr_1d[myps].pstate = (uint8_t)myps;
-		mb_ddr_1d[myps].sequencectrl = sequencectrl[myps];
+		mb_ddr_1d[myps].sequencectrl = (uint16_t)userinputadvanced.sequencectrl[myps];
 		mb_ddr_1d[myps].phyconfigoverride = 0x0U;
 		mb_ddr_1d[myps].hdtctrl = hdtctrl;
 		mb_ddr_1d[myps].msgmisc = msgmisc;
 		mb_ddr_1d[myps].reserved00 = reserved00;
 		mb_ddr_1d[myps].dfimrlmargin = dfimrlmargin;
-		mb_ddr_1d[myps].phyvref = phyvref;
+		mb_ddr_1d[myps].phyvref = (uint8_t)userinputadvanced.phyvref;
 
 #if STM32MP_DDR3_TYPE || STM32MP_DDR4_TYPE
 		mb_ddr_1d[myps].cspresent = cspresent;
-- 
2.53.0

